SDIRK scheme implementation

etiennemarin opened the pull request:
<!-- Please, fill in the description as completely as possible. -->

### Description

This pull request implements the SDIRK (Singly Diagonally Implicit Runge-Kutta) time integration method for the unsteady Navier–Stokes equations.

The current implementation is quite brute-force and would surely benefit from a proper refactoring to improve efficiency and readability. The idea is to also use the SDIRK scheme for other applications beyond just `lethe-fluid`.

The code is structured to match as closely as possible the existing BDF structure. Here's how it works:

- At each time iteration, we loop over the stages. For "traditional" schemes (mainly BDF), the loop is straightforward: there is only one stage.
- At each stage, for the SDIRK scheme, we compute the vectors **kᵢ**.
- However, the structure of Lethe is designed to solve the nonlinear system on the velocity. So instead of directly solving `kᵢ = F(tᵢ, uᵢ)`, we solve the equivalent system:

 `(uᵢ* - uₙ) / (h aᵢᵢ) - ∑_{j=0}^{i-1} (aᵢⱼ / aᵢᵢ) kⱼ = F(tᵢ, uᵢ)`

- Once we have computed the $\mathbf{u}_i^*$ vector, we store the associated $\mathbf{k}_i$ vector in order to compute the next stages and the final sum.
- At the end of the loop, the solution is computed as:

`uₙ₊₁ = uₙ + h ∑ bᵢ kᵢ`

The main changes are implemented in the `solvers/navier_stokes_base.cc` file, within the `iterate()` function.  
The Butcher table is stored in `scratch-data` because it's required during matrix assembly.

---

### Testing

The code has been tested on the `lethe-fluid/application-test` Taylor–Green vortex (GLS formulation) and yields results close to the BDF2 scheme.  
The test outputs, at each time step, the enstrophy, the kinetic energy, and the L2 error.

The goal is now to test this new scheme on multiple examples and with different Butcher tables.

---

### Documentation

- The Butcher table is set in the `scratch-data`.
- The variable `sum_over_previous_stages` is now required for matrix assembly.


### Miscellaneous (will be removed when merged)

<!-- Anything that you would like to add that does not fit into any of the categories above.
       Note that any critical information should be in the categories above.
       Examples:
         Future changes or features that will be added in subsequent pull requests
         Any comments or highlights for the reviewers -->

### Checklist (will be removed when merged)
See [this page](https://chaos-polymtl.github.io/lethe/documentation/contributing.html#pull-requests) for more information about the pull request process.

Code related list:
- [x] All in-code documentation related to this PR is up to date (Doxygen format)
- [x] Copyright headers are present and up to date
- [x] Lethe documentation is up to date
- [x] New feature has unit test(s) (preferred) or application test(s), and restart files are in the generator folder
- [x] The branch is rebased onto master
- [x] Changelog (CHANGELOG.md) is up to date
- [x] Code is indented with indent-all and .prm files (examples and tests) with prm-indent

Pull request related list:
- [x] Labels are applied
- [x] There are at least 2 reviewers (or 1 if small feature) excluding the responsible for the merge
- [x] If this PR closes an issue or is related to a project, it is linked in the "Projects" or "Development" section
- [x] If the fix is temporary, an issue is opened
- [x] The PR description is cleaned and ready for merge

blaisb commented on include/solvers/navier_stokes_scratch_data.h:
```diff
@@ -1086,6 +1101,7 @@ class NavierStokesScratchData : public PhysicsScratchDataBase
   std::vector<Tensor<1, dim>>              pressure_gradients;
   std::vector<std::vector<double>>         previous_pressure_values;
   std::vector<std::vector<Tensor<1, dim>>> previous_velocity_values;
+  std::vector<Tensor<1, dim>>              u_sum_over_stages;
```
I would maybe call it something else than u_sum because it's not the velocity. that's a bit confusing.

etiennemarin replied:
I called it like this because the object FEValues in the scratch_data is associated to the velocity field :
this->fe_values[velocities].get_function_values(sum_over_previous_stages,
                                                    this->u_sum_over_stages);
Is it better to just call it sum_a_ij_k_j and explain what are a_ij coefficients and k_j ?

blaisb commented on source/core/utilities.cc:
```diff
@@ -547,8 +547,7 @@ std::string
 {
-  std::string return_value;
-
+  std::string   return_value;
```
Why is this changed?

etiennemarin commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -553,74 +565,227 @@ template <int dim, typename VectorType, typename DofsType>
+                      tmp                      = previous_k_j_solutions[0];
+                      tmp.equ(stage_data.a_ij[p] / a_ii,
+                              previous_k_j_solutions_p);
+                      temp_sum_over_previous_stages.add(1.0, tmp);
```
Yes, I don't need to have all these lines, your solution works. But we still need to have a non-locally relevant vector in the .add operation, and previous_k_j_solutions[p] is locally-relevant

etiennemarin commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -553,74 +565,227 @@ template <int dim, typename VectorType, typename DofsType>
+                      previous_k_j_solutions_p = previous_k_j_solutions[p];
+                      tmp                      = previous_k_j_solutions[0];
+                      tmp.equ(stage_data.a_ij[p] / a_ii,
+                              previous_k_j_solutions_p);
```
In the operations (addition, multiplication, etc.) we need a non-locally relevant dofs vector, I think

brunaccampos reviewed (changes requested):
Some comments and some questions :) nice work

etiennemarin commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -567,81 +572,242 @@ NavierStokesBase<dim, VectorType, DofsType>::finish_time_step()
+  // At each time iteration, we reset the value of sum_bi_ki
+  temp_sum_bi_ki = 0;
+
+  for (unsigned int stage = 0; stage < n_stages; ++stage)
```
Yes, for bdf, n_stages is initialized as one, so the loop has only one iteratation.

etiennemarin commented on source/core/sdirk_stage_data.cc:
```diff
@@ -66,12 +66,50 @@ sdirk_table(const Parameters::SimulationControl::TimeSteppingMethod method)
       table.A(0, 0) = gamma;
```
Yes they are fixed, I'll add that on the description

etiennemarin commented on include/solvers/navier_stokes_assemblers.h:
```diff
@@ -308,6 +308,54 @@ class GLSNavierStokesAssemblerBDF : public NavierStokesAssemblerBase<dim>
+ * stages,
+ * - \( h \) is the time step size,
+ * - \( a_{ij} \), \( a_{ii} \) are coefficients from the SDIRK Butcher tableau.
+ *
```
Right, I'll changed that :)

etiennemarin commented on doc/source/parameters/cfd/simulation_control.rst:
```diff
@@ -107,6 +107,9 @@ This subsection is the most important in a simulation and therefore, the most co
 	* ``bdf1``: 1st order backward differentiation
 	* ``bdf2``: 2nd order backward differentiation
 	* ``bdf3``: 3rd order backward differentiation
+	* ``sdirk22``: 2nd order 2 stages singly diagonally implicit Runge-Kutta
```
Exactly, the only thing to change in the .prm file is the 'method = ' section. I'll add a note

etiennemarin commented on include/solvers/navier_stokes_scratch_data.h:
```diff
@@ -100,6 +101,15 @@ class NavierStokesScratchData : public PhysicsScratchDataBase
                      update_values | update_quadrature_points |
                        update_JxW_values | update_gradients | update_hessians |
                        update_normal_vectors)
+    , sdirk_table(
```
Yes, it is initialized as empty if a BDF or steady method is chosen. I would say that it doesn't take much computional cost, so I Implemented like this. But I don't know if there is an easier way to implement this

blaisb replied:
No this is fine. But instead of manually checking for every SDIRK method, you should call your is_sdirk() method

etiennemarin commented on applications_tests/lethe-fluid/taylor-green-vortex_gls_sdirk33.prm:
Actually the SDIRK43 doesn't reach order 4 as expected (approximately 3.5) so I don't know if, in the end, this scheme will be kept in the code...

etiennemarin commented on applications_tests/lethe-fluid/taylor-green-vortex_gls_sdirk33.prm:
```diff
@@ -0,0 +1,151 @@
+
+subsection simulation control
+  set method            = sdirk33
+  set time step         = 0.2 # Time step
```
You're right, I will harmonize this :)

etiennemarin commented on source/solvers/fluid_dynamics_matrix_based.cc:
```diff
@@ -103,10 +109,37 @@ FluidDynamicsMatrixBased<dim>::setup_dofs_fd()
+
+  this->sdirk_vectors.locally_owned_for_calculus.reinit(
+    this->locally_owned_dofs, this->mpi_communicator);
+
```
If I don't reinit them even if I don't use them, I have this assertion when the tests are running :
The violated condition was: 
    values.size() == this->get_dof_handler().n_dofs()
For now, I just add a comment and remove the if condition

blaisb commented on include/solvers/navier_stokes_scratch_data.h:
```diff
@@ -268,6 +280,11 @@ class NavierStokesScratchData : public PhysicsScratchDataBase
       compute_cell_measure_with_JxW(this->fe_values.get_JxW_values());
     this->cell_size = compute_cell_diameter<dim>(cell_measure, fe.degree);
 
+    // For the SDIRK methods, \sum_{j=1}^{i-1} a_{ij} k_j is needed for the
+    // assembler
+    this->fe_values[velocities].get_function_values(sum_over_previous_stages,
+                                                    this->sdirk_stage_sum);
```
You have access to the simulation control right? I wonder if  we could get_function_values of this field only and only if the method is sdirk? This would save additional calculations.

blaisb commented on source/solvers/fluid_dynamics_matrix_based.cc:
```diff
@@ -103,10 +109,36 @@ FluidDynamicsMatrixBased<dim>::setup_dofs_fd()
                       this->mpi_communicator);
     }
 
+  // Reinitialize vectors used for the SDIRK methods
+  this->sdirk_vectors.sum_bi_ki.reinit(this->locally_owned_dofs,
+                                       this->locally_relevant_dofs,
+                                       this->mpi_communicator);
+  this->sdirk_vectors.temp_sum_bi_ki.reinit(this->locally_owned_dofs,
+                                            this->mpi_communicator);
+
+
+  this->sdirk_vectors.sum_over_previous_stages.reinit(
+    this->locally_owned_dofs,
+    this->locally_relevant_dofs,
+    this->mpi_communicator);
+  this->sdirk_vectors.temp_sum_over_previous_stages.reinit(
+    this->locally_owned_dofs, this->mpi_communicator);
+
+  this->sdirk_vectors.locally_owned_for_calculus.reinit(
+    this->locally_owned_dofs, this->mpi_communicator);
```
I would reinitialize these vector only if the method is sdirk. you have this nice function in your simulation control to check it :)
Will save calculations for the rest of simulations.

blaisb commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -567,82 +572,242 @@ NavierStokesBase<dim, VectorType, DofsType>::finish_time_step()
+  auto &temp_sum_over_previous_stages =
+    this->sdirk_vectors.temp_sum_over_previous_stages;
+
+  auto &locally_owned_for_calculus =
+    this->sdirk_vectors.locally_owned_for_calculus;
```
calculus -> calculation

blaisb commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -567,82 +572,242 @@ NavierStokesBase<dim, VectorType, DofsType>::finish_time_step()
+  unsigned int                                      stage,
+  Parameters::SimulationControl::TimeSteppingMethod method)
+{
+  // Very important variable because sum(a_{ij} * k_j) has to be
+  // calculated at each stage
+  auto &previous_k_j_solutions = this->sdirk_vectors.previous_k_j_solutions;
+
+  // sum(a_{ij} *k_j) when we are at least at stage 2
+  // If we are at stage 1, sum_over_previous_stages is set to 0.
+  auto &sum_over_previous_stages = this->sdirk_vectors.sum_over_previous_stages;
+  auto &temp_sum_over_previous_stages =
+    this->sdirk_vectors.temp_sum_over_previous_stages;
+
+  auto &locally_owned_for_calculus =
```
Here I would add a comment that first you copy the reference to some of the vectors to enhance readability below. I have nothing agasint that, it is more readable, but I would explain that more clearly :)

blaisb commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -567,82 +572,242 @@ NavierStokesBase<dim, VectorType, DofsType>::finish_time_step()
 NavierStokesBase<dim, VectorType, DofsType>::iterate()
 {
+  // The locally_relevant_dofs are used for the assembly,
+  // while the non-locally_relevant_dofs are used for the calculus on the
+  // vectors (addition, multiplication, etc.).
```
```suggestion
  // while the non-locally_relevant_dofs are used for the calculations on the
  // vectors (addition, multiplication, etc.).
```

blaisb commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -567,82 +572,242 @@ NavierStokesBase<dim, VectorType, DofsType>::finish_time_step()
-      multiphysics->percolate_time_vectors(true);
-    }
-  else
+  // The following variables are used for the SDIRK method
+  // u_{n+1} = u_n + sum(b_i * k_i)
+  auto &temp_sum_bi_ki = this->sdirk_vectors.temp_sum_bi_ki;
+  // At each time iteration, we reset the value of sum_bi_ki
+  temp_sum_bi_ki = 0;
```
I would do this only if this is an sdirk method. You don,t need these tables and operation if it's not sdirk :).

blaisb commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -567,82 +572,242 @@ NavierStokesBase<dim, VectorType, DofsType>::finish_time_step()
-  else
+  // The following variables are used for the SDIRK method
+  // u_{n+1} = u_n + sum(b_i * k_i)
+  auto &temp_sum_bi_ki = this->sdirk_vectors.temp_sum_bi_ki;
```
Here I would avoid creating this temporary variable and just use the sdirk_vectors.temp_sum_bi_ki variable directly down there.

blaisb commented on source/solvers/navier_stokes_scratch_data.cc:
```diff
@@ -39,6 +39,8 @@ NavierStokesScratchData<dim>::allocate()
   this->velocity_hessians    = std::vector<Tensor<3, dim>>(n_q_points);
   this->shear_rate           = std::vector<double>(n_q_points);
 
+  this->sdirk_stage_sum = std::vector<Tensor<1, dim>>(n_q_points);
```
I would consider only allocating if the method is sdirk

blaisb reviewed (changes requested):
A few last comments. Very good PR and very good work this was not easy.
Your implementation is 10X better than the one I had done in the past. It's a lot more reusable.

OGaboriault commented on applications_tests/lethe-fluid/taylor-green-vortex_gls_sdirk22.prm:
```diff
@@ -0,0 +1,151 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 The Lethe Authors
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR LGPL-2.1-or-later
+
```
Can you add a comment in the prm to know what this is testing ?

OGaboriault commented on include/core/parameters.h:
```diff
@@ -81,7 +81,8 @@ namespace Parameters
       sdirk22,
-      sdirk33
+      sdirk33,
+      sdirk43
```
do we need the second number? Can't this just be sdirk2, sdirk3 and sdirk4 ?

etiennemarin replied:
There are methods of order 2 with 3 stages, or order 4 with 4 stages for example that are not yet implemented here. For now, I don't really know what are the most relevant Butcher's tables to use for the simulations, but maybe we will need all of these in the code. With this nomenclature we can have all of them :)

OGaboriault commented on include/solvers/navier_stokes_base.h:
```diff
@@ -358,6 +359,42 @@ class NavierStokesBase : public PhysicsSolver<VectorType>
   virtual void
   dynamic_flow_control();
 
+  /**
+   * @brief multi_stage_preresolution
+   * It updates the sum_over_previous_stages variable to use it in the solver.
```
```suggestion
  /**
   * @brief Update the sum_over_previous_stages variable to used in the solver.
```

OGaboriault commented on include/solvers/navier_stokes_base.h:
```diff
@@ -358,6 +359,42 @@ class NavierStokesBase : public PhysicsSolver<VectorType>
+    Parameters::SimulationControl::TimeSteppingMethod method);
+
+  /**
+   * @brief multi_stage_postresolution
```
Don't write the name of the function in the brief, just the description of what it is doing.

OGaboriault replied:
I think the file was using this convention before your PR, but other files are not. Could you fix this problem for every function in this file?

OGaboriault commented on source/core/parameters.cc:
```diff
@@ -78,11 +78,13 @@ namespace Parameters
+        "steady",
+        Patterns::Selection(
+          "steady|steady_bdf|bdf1|bdf2|bdf3|sdirk22|sdirk33|sdirk43"),
+        "The kind of solver for the linear system. "
```
Is this really a kind of solver?

etiennemarin replied:
I don't know. It was already there, but actually I don't really know what does it mean. Maybe "method" is more accurate @blaisb ?

OGaboriault commented on source/core/sdirk_stage_data.cc:
```diff
@@ -50,28 +54,75 @@ sdirk_table(const Parameters::SimulationControl::TimeSteppingMethod method)
-  // Important note : the nomenclature used for the name of the SDIRK methods
-  // are sdirkOrderStage sdirk22 means SDIRK with order 2 and 2 stages, sdirk33
-  // means SDIRK with order 3 and 3 stages
+  // Important note : When referencing to SDIRK methods the pattern used is
```
I think you can remove this comment since it is written at line 27

OGaboriault commented on source/core/sdirk_stage_data.cc:
```diff
@@ -50,28 +54,75 @@ sdirk_table(const Parameters::SimulationControl::TimeSteppingMethod method)
+    }
+
+
+  // Important note : When referencing to SDIRK methods the pattern used is
```
Same here

OGaboriault commented on source/solvers/fluid_dynamics_matrix_based.cc:
```diff
@@ -87,10 +87,16 @@ FluidDynamicsMatrixBased<dim>::setup_dofs_fd()
   this->reinit_mortar();
 
+
+  // Important note : operations on the following vectors (addition,
```
You can remove the "Important note :" part from this comment. 

Every comment is an important note ;)

OGaboriault commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -567,82 +572,242 @@ NavierStokesBase<dim, VectorType, DofsType>::finish_time_step()
+  unsigned int                                      stage,
+  Parameters::SimulationControl::TimeSteppingMethod method)
+{
+  // Very important variable because sum(a_{ij} * k_j) has to be
+  // calculated at each stage
+  auto &previous_k_j_solutions = this->sdirk_vectors.previous_k_j_solutions;
```
I would reformulate this comment. 

something like : "sum(a_{ij} * k_j) has to be calculated at each stage. "

etiennemarin commented:
> Here are my comments!
> 
> I have just one question, if we are also setting VOF = true or any other physics, which scheme is taken? Or do we have a flag already since SDIRK is only implemented for NS?

etiennemarin commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -172,6 +172,10 @@ NavierStokesBase<dim, VectorType, DofsType>::NavierStokesBase(
   // of the BDF schemes
   previous_solutions.resize(maximum_number_of_previous_solutions());
```
We need the previous_solutions[0] vector only, not the others !

etiennemarin replied:
Maybe I can resize it with only one element ?

etiennemarin commented:
I corrected with all your advices. Thank you all for your time :)). For the comment of Gabo to harmonize the "@brief" in the navier_stokes_base.h file, I removed all the name of the functions from the brief, because I think it is already written by Dioxygen

blaisb commented on source/solvers/fluid_dynamics_block.cc:
```diff
@@ -602,10 +604,15 @@ FluidDynamicsBlock<dim>::setup_dofs_fd()
   this->define_zero_constraints();
 
+  // Important note : operations on the following vectors (addition,
+  // multiplication, etc.) can only be done if these are reinitialized WITHOUT
+  // locally_relevant_dofs. This is why most of them are reinitialized both with
```
@etiennemarin I would add this suggestion

blaisb commented on source/solvers/navier_stokes_base.cc:
```diff
@@ -168,9 +168,20 @@ NavierStokesBase<dim, VectorType, DofsType>::NavierStokesBase(
+    {
+      // Pre-allocate memory for the previous solutions using the information
+      // of the SDIRK schemes
+      previous_solutions.resize(1);
```
```suggestion
```
Let's wait a little for this optimization :) This can come afterwards ok? :)

blaisb reviewed (approved):
3 last things to address, then this will be good for merge in my opinion.
Please wait for the ok from @hepap and @brunaccampos  before marking reviewed and ready to merge please :)

blaisb commented on source/solvers/fluid_dynamics_matrix_based.cc:
Instead of putting the assertions in navier_stokes_base, could you please put them in their respective physics (like in the constructor)
Just put these assertion in the constructor of:
- The tracer
- The heat transfer
- VOF
- CHN

:)

blaisb reviewed (approved):
Great work @etiennemarin . I know this is a lot of work, but it's normal :). Keep it up and one more day and it will be merged I think :)

blaisb commented:
@etiennemarin One last thing to check with one of  your test, then this is good for a mergggge :)
